### PR TITLE
Monthly Iterator

### PR DESCRIPTION
**Pull request by bmuniswamy**
# Critical Changes
    
# Changes
 Added the MonthlyIterator class and test class. Modified the RecurrenceRule,RecurrenceFactory and RecurrenceService classes to add monthly functionality. Added tests to all the Recurrence classes.
# Issues Closed

# New Metadata

# Deleted Metadata

# Definition of Done
  Refer to [Asteroids DoD document](https://salesforce.quip.com/iq2mAy4i62oM) to see any additional details for the items below
~~- [ ] Place holder data is incorporated (Refer to [Case Management Sample Data document](https://quip.com/cbFcAPJF8t0z))~~
~~- [ ] Base axe-core JEST test included for any new LWC (No critical violations)~~
~~- [ ] CRUD/FLS is enforced in Apex~~
~~- [ ] Update & Test permission sets to account for new features and CRUD/FLS~~
- [x] All modified classes are unit tested (Apex) at 100% coverage
~~- [ ] UX approval or UX not necessary~~
- [x] PR contains draft release notes
- [x] Attach work item to the pull request with [Lurch](https://salesforce.quip.com/50ZRA5LEWVzH) `**Lurch:attach W-000000` or `**Lurch:add`
- [ ] All **acceptance criteria** have been met
    - [x] Developer
    - [ ] Code Reviewer
    - [ ] QA
- [ ] QE story level testing completed

### CODE REVIEW

**Comment by bmuniswamy** [comment]
[W-8043620](https://gus.lightning.force.com/lightning/n/ADM_Work_Backlog)

File: force-app/main/default/classes/RecurrenceRule_TEST.cls [hunk 1]
```diff
@@ -84,6 +84,50 @@ public with sharing class RecurrenceRule_TEST {
         System.assertEquals(days, rule.getDays());
     }
 
+    @IsTest
+    private static void shouldSetAndGetBySetPos() {
+        Integer bySetPos = -1;
```
Comment (bethbrains) [inline force-app/main/default/classes/RecurrenceRule_TEST.cls:89]: Can we call this defaultBySetPos in the test to make it clear we're testing a default? Took me a sec to remember why there was no `.with`... here. :)

**Review by bethbrains — CHANGES_REQUESTED** [review]
This looks great! A few nitpicky naming things, but no big deals at all. Great work! Very clever.

File: force-app/main/default/classes/MonthlyIterator.cls [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    public Integer getMaxByMonthDays(Set<Integer> byMonthDays) {
+        List<Integer> days = new List<Integer>();
+        Integer maxValue = 0;
+        Integer count = 0;
+
+        //adding the values of the set so we can get the max value from the list
+        days.addAll(byMonthDays);
+
+        for (Integer i = 0; i < days.size(); i++) {
```
[the remark this answers is not in the document]
Comment (bmuniswamy) [reply force-app/main/default/classes/MonthlyIterator.cls:69]: We could use that.

**Comment by bmuniswamy** [comment]
@bethbrains , I have made the necessary changes that you had requested.

**Review by bethbrains — APPROVED** [review]
Awesome work!!

File: force-app/main/default/classes/MonthlyIterator.cls [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        do {
+            if (rRule.getBySetPos() != null && rRule.getByMonthDays() != null) {
+                //The startDay variable will hold the max value that is returned by rRule.getByMonthDays.
+                startDay = getMaxByMonthDays(rRule.getByMonthDays());
+                result.add(DateTime.newInstance(runningDate, dateTimeStart.time()));
+
+                //Here we are adding days after adding months. So the number of days we are adding will
+                //be the difference of startDay and runningDate.day().
+
+                //Eg : if startdate is 1/31/2020 and we are creating occurences until 04/30/2020. The start date is added
```
Comment (jjbennett) [inline force-app/main/default/classes/MonthlyIterator.cls:25]: Super smart, there is still something bugging me at the back of my mind. Can you add a test that starts on 12/31 and goes every other month?

File: force-app/main/default/classes/MonthlyIterator_TEST.cls [hunk 4]
```diff
@@ -0,0 +1,416 @@
+@IsTest
```
Comment (jjbennett) [inline force-app/main/default/classes/MonthlyIterator_TEST.cls:1]: I love that we have so many tests but I'm wondering if we should use a more varied start date, or add additional tests with different start dates.

File: force-app/main/default/classes/RecurrenceRule_TEST.cls [hunk 5]
```diff
@@ -84,6 +84,50 @@ public with sharing class RecurrenceRule_TEST {
         System.assertEquals(days, rule.getDays());
     }
 
+    @IsTest
+    private static void shouldSetAndGetBySetPos() {
```
Comment (jjbennett) [inline force-app/main/default/classes/RecurrenceRule_TEST.cls:88]: Should this be named should get Default vs shouldSet?

File: force-app/main/default/classes/RecurrenceRule_TEST.cls [hunk 6]
```diff
@@ -210,4 +255,34 @@ public with sharing class RecurrenceRule_TEST {
...
+        RecurrenceRule rule = new RecurrenceRule()
+            .withFrequency('MONTHLY')
+            .withEndDate(Date.newInstance(2020, 7, 31))
+            .withByMonthDays(DateTime.newInstance(2020, 1, 31, 13, 0, 0));
+
+        Test.startTest();
+        String actual = rule.build();
+        Test.stopTest();
+
+        System.assertEquals(expected, actual);
```
Comment (jjbennett) [inline force-app/main/default/classes/RecurrenceRule_TEST.cls:272]: Do we want to add messages to all of the asserts like we talked about the other day?

**Review by jjbennett — COMMENTED** [review]
Awesome work, I had a couple minor comments and then one test I was hoping you could add, please let me know if you would like to discuss.

File: force-app/main/default/classes/MonthlyIterator.cls (hunk 3 above)
Comment (bmuniswamy) [reply force-app/main/default/classes/MonthlyIterator.cls:25]: @jjbennett , There is a test that tests this scenario. The name of the test method is `shouldIterateEveryLastDayOfTheMonthUntilGivenEndDateOverLeapYearWithIntervalOf2`. Let me know if you want me to add additional tests

File: force-app/main/default/classes/MonthlyIterator_TEST.cls (hunk 4 above)
Comment (bmuniswamy) [reply force-app/main/default/classes/MonthlyIterator_TEST.cls:1]: sure

**Review by bethbrains — APPROVED** [review]
Great work! MVP ho!!

**Pull request merged**